services/merak-topo/handler/handler.go

```diff
@@ -15,70 +15,89 @@ package handler
-	log.Printf("Vhost number is: %v\n", aca_num)
-	log.Printf("Rack number is: %v\n", rack_num)
-	log.Printf("Vhosts per rack is: %v\n", aca_per_rack)
 
-	log.Printf("Ports per vswitch is: %v\n", ports_per_vswitch)
+	var aca_image string
+	var ovs_image string
+	var err_return error
+
+	err_flag := 0
+	errs := errors.New("request DEPLOY- fails to create topology")
+
+	for _, img := range images {
+		if strings.Contains(img.Name, "ACA") {
+			aca_image = img.Registry
+		} else if strings.Contains(img.Name, "OVS") {
+			ovs_image = img.Registry
+		}
+	}
 
-	log.Println("CREATE:====== Start: Generate multi-layer topology structure ======= ")
+	utils.Logger.Debug("request DEPLOY details", "Vhost number", aca_num, "Rack number", rack_num, "Vhosts per rack", aca_per_rack, "Ports per vswitch", ports_per_vswitch)
 
-	err_create, topo := Create_multiple_layers_vswitches(int(aca_num), int(rack_num), int(aca_per_rack), int(ports_per_vswitch), data_plane_cidr)
+	topo, err_create := Create_multiple_layers_vswitches(int(aca_num), int(rack_num), int(aca_per_rack), int(ports_per_vswitch), data_plane_cidr)
 	if err_create != nil {
-		return fmt.Errorf("create multiple layers vswitches error %s", err_create)
+		utils.Logger.Error("request DEPLOY", "multiple layers vswitches", err_create.Error())
+		returnMessage.ReturnMessage = "Can not create multiple layers vswitches"
+		returnMessage.ReturnCode = pb_common.ReturnCode_FAILED
+		err_flag = 1
+
 	}
 
 	topo.Topology_id = topo_id
 
 	elaps0 := time.Since(start_time)
 	start0 := time.Now()
 
-	log.Printf("CREATE:=== Complete: Generate topology data === %v", elaps0)
+	utils.Logger.Info("request DEPLOY", " Complete: Generate topology data (in second)", elaps0)
 
-	err1 := Topo_save(topo)
-	if err1 != nil {
-		return fmt.Errorf("save topology to redis error %s", err1)
+	err_db := database.SetValue(topoPrefix, topo)
+	if err_db != nil {
+		utils.Logger.Error("request DEPLOY", "save topology to redis", err_db.Error(), "topo_id", topoPrefix)
+		returnMessage.ReturnMessage = "Can not save topology to redis"
+		returnMessage.ReturnCode = pb_common.ReturnCode_FAILED
+		err_flag = 1
 	}
 
 	elaps1 := time.Since(start0)
 	start1 := time.Now()
-	log.Printf("CREATE:=== Complete: Save topology to redis DB === %v", elaps1)
+	utils.Logger.Info("request DEPLOY", "Save topology in redis DB (in second)", elaps1)
 
-	k8s_nodes, err1 := k8client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	k8s_nodes, err2 := k8client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
 
-	if err1 != nil {
-		return fmt.Errorf("CREATE: failed to list k8s host nodes info %s", err1)
+	if err2 != nil {
+		utils.Logger.Error("request DEPLOY", "check nodes in k8s cluster", err2.Error())
+		returnMessage.ReturnMessage = "CREATE: can not list k8s host nodes info"
+		returnMessage.ReturnCode = pb_common.ReturnCode_FAILED
+		err_flag = 1
 	}
 
 	for _, s := range k8s_nodes.Items {
 		var hnode database.HostNode
 
```

Comment:
@zzengcs if `err_flag=1`, that mean k8s create nodes failed and `k8s_nodes` will be empty. Then the next `for` loop will get nil pointer for `k8s_nodes.Items`.